### PR TITLE
perf(@angular-devkit/build-angular): inject Sass import/use directive importer information when resolving

### PR DESCRIPTION
**Pull request by clydin**
To correctly resolve a package based import reference in a Sass file with pnpm or Yarn PnP, the importer file path must be known. Unfortunately, the Sass compiler does not provided the importer file to import plugins. Previously to workaround this issue, all previously resolved stylesheets were tried as the importer path. This allowed the stylesheets to be resolved but it also could cause a potentially large increase in build time due to the amount of previous stylesheets that would need to be tried. To avoid the performance impact and to also provide more accurate information regarding the importer file, a lexer is now used to extract import information for a stylesheet and inject the importer file path into the specifier. This information is then extracted from the import specifier during the Sass resolution process and allows the underlying package resolution access to a viable location to resolve the package for all package managers. This information is currently limited to specifiers referencing the `@angular` and `@material` package scopes but a comprehensive pre-resolution process may be added in the future.

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 1]
```diff
@@ -74,6 +83,26 @@ abstract class UrlRebasingImporter implements Importer<'sync'> {
       updatedContents.update(start, end, rebasedUrl);
     }
 
+    for (const { start, end, value } of findImports(contents)) {
+      // Currently only provide directory information for known/common packages:
+      // * `@material/`
+      // * `@angular/`
+      // Comprehensive pre-resolution support may be added in the future.
+      if (!value.startsWith('@angular/') && !value.startsWith('@material/')) {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:91]: Question: Any particular reason for limiting to `@angular/` and `@material/`? Just staying safe in case it breaks something?

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 2]
```diff
@@ -74,6 +83,26 @@ abstract class UrlRebasingImporter implements Importer<'sync'> {
       updatedContents.update(start, end, rebasedUrl);
     }
 
+    for (const { start, end, value } of findImports(contents)) {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:86]: Suggestion: Rename `value` to something more meaningful. Maybe `importSpecifier`?

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 3]
```diff
@@ -444,20 +315,41 @@ export class ModuleUrlRebasingImporter extends RelativeUrlRebasingImporter {
...
+      specifier: string,
+      options: { fromImport: boolean; resolveDir?: string },
+    ) => URL | null,
   ) {
     super(entryDirectory, directoryCache, rebaseSourceMaps);
   }
 
   override canonicalize(url: string, options: { fromImport: boolean }): URL | null {
-    if (url.startsWith('file://')) {
+    if (url[0] === 'f' && url.startsWith('file://')) {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:329]: Question: What's the need for `url[0] === 'f'`? Is that a performance optimization? Is `startsWith` that slow that it makes a difference?

I also see this pattern used a couple times. Should we make a separate `fastStartsWith` function which reads `substring[0]` so we don't have to duplicate the first character and can comment explaining why it's necessary to do this?

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 4]
```diff
@@ -444,20 +315,41 @@ export class ModuleUrlRebasingImporter extends RelativeUrlRebasingImporter {
...
 
-    const result = this.finder(url, options);
+    let resolveDir;
+    if (url[0] === 'n' && url.startsWith('ng-package;')) {
+      const values = url.split(';', 3);
+      resolveDir = values[1];
+      url = values[2];
+    }
+
+    let result: URL | undefined | null = this.moduleCache.get(url)?.resolved;
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:340]: The module cache seems kind of unrelated to this rebaser/canonicalizer. I wonder if we could pull that out to a slightly higher level of abstraction and maybe cache before/after calling `canonicalize()` to separate concerns a little cleaner?

That would probably apply the cache to more than just `ng-package;` specifiers, but maybe that's a good thing? Not sure how viable this is, just a thought which stood out to me.

File: packages/angular_devkit/build_angular/src/tools/sass/lexer.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
+/**
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/lexer.ts:1]: I'm mostly ignoring the lexer since it doesn't look like it changed much. If there's anything in particular I should be looking at, let me know. I recommend moving this refactor into a different commit so the changes to the lexer which support this Sass change show up in the diff more clearly.

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 6]
```diff
@@ -74,6 +83,26 @@ abstract class UrlRebasingImporter implements Importer<'sync'> {
...
+      if (!value.startsWith('@angular/') && !value.startsWith('@material/')) {
+        continue;
+      }
+
+      const resolvePath = stylesheetDirectory + ';' + value;
+
+      // Normalize path separators and escape characters
+      // https://developer.mozilla.org/en-US/docs/Web/CSS/url#syntax
+      const resolvedUrl =
+        'ng-package;' + resolvePath.replace(/\\/g, '/').replace(/[()\s'"]/g, '\\$&');
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:100]: Question: Would this break if the user named a file/project `ng-package`? I know it's unlikely, but not exactly implausible. Would something like `__ng-package__` be a little safer?

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 7]
```diff
@@ -444,20 +315,41 @@ export class ModuleUrlRebasingImporter extends RelativeUrlRebasingImporter {
...
 
   override canonicalize(url: string, options: { fromImport: boolean }): URL | null {
-    if (url.startsWith('file://')) {
+    if (url[0] === 'f' && url.startsWith('file://')) {
       return super.canonicalize(url, options);
     }
 
-    const result = this.finder(url, options);
+    let resolveDir;
+    if (url[0] === 'n' && url.startsWith('ng-package;')) {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:334]: The two `ng-package` usages, where we create the fake module specifier and then unpack it later, are kind of far apart and not clearly connected.

Suggestion: Make a pair of functions to pack/unpack the data in this string format, possibly with a constant for `ng-package` to be clear that we're defining a special string format and using it in two places which are tied together. This will make it more obvious that updating one also requires updating the other.

**Review by dgp1130 — COMMENTED** [review]
Do we have any existing unit tests for this functionality we should be updating as well? Or is the existing integration test suite sufficient for this change, given that it's a performance optimization which shouldn't observably change anything?

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts (hunk 3 above)
Comment (clydin) [reply packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:329]: it is a micro-optimization. dropped it for now and can re-evaluate in the future with more profiling.

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts (hunk 4 above)
Comment (clydin) [reply packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:340]: Currently it would apply to all non-relative resolution attempts.  Since we don't control the execution of `canonicalize` it needs to be done inside. `canonicalize` is called by the Sass compiler itself and is akin to a onResolve hook.

I dropped the cache from this PR.  It does result in less of a performance improvement overall here but it can be introduced in a separate change. I wanted to eventually make the cache work across all active Sass workers anyway.

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 8]
```diff
@@ -22,6 +23,38 @@ export interface DirectoryEntry {
...
+/**
+ * A prefix that is added to import and use directive specifiers that should be resolved
+ * as modules and that will contain added resolve directory information.
+ *
+ * This functionality is used to workaround the Sass limitation that it does not provide the
+ * importer file to custom resolution plugins.
+ */
+const MODULE_RESOLUTION_PREFIX = '__NG_PACKAGE__';
+
+function packModuleSpecifier(specifier: string, resolveDir: string): string {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:35]: Suggestion: An interface for the unpacked format would allow you to accept/return the same type rather than disparite values.

```typescript
interface ModuleSpecifier {
  specifier: string;
  resolveDir?: string;
}
```

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts [hunk 9]
```diff
@@ -22,6 +23,38 @@ export interface DirectoryEntry {
...
+
+  // Normalize path separators and escape characters
+  // https://developer.mozilla.org/en-US/docs/Web/CSS/url#syntax
+  const normalizedPacked = packed.replace(/\\/g, '/').replace(/[()\s'"]/g, '\\$&');
+
+  return normalizedPacked;
+}
+
+function unpackModuleSpecifier(specifier: string): { specifier: string; resolveDir?: string } {
+  if (specifier.startsWith(`${MODULE_RESOLUTION_PREFIX};`)) {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts:46]: Nit: The "interesting" part of the function is the `specifier.split(';', 3)`. I suggest inverting the `if` statement to early return `{ specifier }` in the uninteresting case so the rest of the function can focus on the interesting one.

**Review by dgp1130 — APPROVED** [review]

File: packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts (hunk 1 above)
Comment (clydin) [reply packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts]: Safer for now and avoids some more complex logic for determining if it is relative or an actual module reference.  CSS/Sass don't require `./` or `../` for relative paths. So for general support, a bare specifier would need to be checked for relative path viability first. Technically a project could have a relative `@angular` directory but that's unlikely and easy to check/remedy (add `./` to the specifier).

Comment (dgp1130) [reply packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts]: Makes sense. Might be useful to include some of that in the comment so we know the constraints if/when we change or open up the package restriction.
